File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/db/chat/ChatDB.kt
package jp.co.toukei.log.trustar.db.chat

import androidx.room.Database
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import jp.co.toukei.log.lib.room.Converter
import jp.co.toukei.log.trustar.db.chat.entity.ChatMessage
import jp.co.toukei.log.trustar.db.chat.entity.ChatMessagePending
import jp.co.toukei.log.trustar.db.chat.entity.ChatRoom
import jp.co.toukei.log.trustar.db.chat.entity.ChatRoomUser
import jp.co.toukei.log.trustar.db.chat.entity.ChatUser

@Database(
    entities = [
        ChatRoom::class,
        ChatUser::class,
        ChatRoomUser::class,
        ChatMessage::class,
        ChatMessagePending::class
    ],
    exportSchema = false,
    version = 2
)
@TypeConverters(Converter::class)
abstract class ChatDB : RoomDatabase() {

    abstract fun chatRoomDao(): ChatRoomDao
    abstract fun chatUserDao(): ChatUserDao
    abstract fun chatRoomUserDao(): ChatRoomUserDao
    abstract fun chatMessageDao(): ChatMessageDao
    abstract fun chatMessagePendingDao(): ChatMessagePendingDao
}

File: LogiOS/TrustarApp/Database/chat/ChatDb.swift
//
//  ChatDb.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/11/24.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation
import GRDB

class ChatDb: BaseDb {
    
    var chatRoomDao: ChatRoomDao? = nil
    var chatUserDao: ChatUserDao? = nil
    var chatRoomUserDao: ChatRoomUserDao? = nil
    var chatMessageDao: ChatMessageDao? = nil
    var chatMessagePendingDao: ChatMessagePendingDao? = nil
        
    init(_ userName: String) throws {
        super.init()
        grdb = GRDBTrustar("chatDb_" + userName)

        let isDbExisted = grdb!.isDbExisted()
        
        instanceDb = try grdb!.makeDatabaseQueue()
        
        if !isDbExisted {
            initTable(table: ChatRoom(), columns: ChatRoomColumns.allCases.map({($0.rawValue, $0.name())}))
            
            initTable(table: ChatUser(), columns: ChatUserColumns.allCases.map({($0.rawValue, $0.name())}))
            
            initTable(table: ChatRoomUser(), columns: ChatRoomUserColumns.allCases.map({($0.rawValue, $0.name())}))
            
            initTable(table: ChatMessage(), columns: ChatMessageColumns.allCases.map({($0.rawValue, $0.name())}))
            
            initTable(table: ChatMessagePending(), columns: ChatMessagePendingColumns.allCases.map({($0.rawValue, $0.name())}))
        }
        
        initDao()
    }

    private func initDao() {
        chatRoomDao = ChatRoomDao(self)
        chatUserDao = ChatUserDao(self)
        chatRoomUserDao = ChatRoomUserDao(self)
        chatMessageDao = ChatMessageDao(self)
        chatMessagePendingDao = ChatMessagePendingDao(self)
    }
}
